File: get_maya_plug4/core/logger.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
日志管理模块
提供统一的日志输出功能，支持输出到控制台和/或文件
"""

import os
import sys
from datetime import datetime
from typing import Optional
from enum import Enum


class LogLevel(Enum):
    """日志级别"""
    DEBUG = 0
    INFO = 1
    WARNING = 2
    ERROR = 3


class Logger:
    """日志管理器"""
    
    def __init__(
        self,
        log_file: Optional[str] = None,
        console_output: bool = True,
        file_output: bool = False,
        log_level: LogLevel = LogLevel.INFO,
        append_mode: bool = True
    ):
        """
        初始化日志管理器
        
        Args:
            log_file: 日志文件路径（如果为None且file_output=True，则自动生成）
            console_output: 是否输出到控制台
            file_output: 是否输出到文件
            log_level: 日志级别（只输出大于等于此级别的日志）
            append_mode: 文件追加模式（True=追加，False=覆盖）
        """
        self.console_output = console_output
        self.file_output = file_output
        self.log_level = log_level
        self.append_mode = append_mode
        
        # 日志目录及文件状态
        self._log_dir: Optional[str] = None
        self._current_day: Optional[str] = None
        self.log_file: Optional[str] = None

        if self.file_output:
            self._configure_log_directory(log_file)
    
    def _format_message(self, message: str, level: LogLevel) -> str:
        """格式化日志消息"""
        timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        level_name = level.name
        return f"[{timestamp}] [{level_name}] {message}"
    
    def _should_log(self, level: LogLevel) -> bool:
        """判断是否应该输出此级别的日志"""
        return level.value >= self.log_level.value
    
    def _write_to_console(self, formatted_message: str) -> None:
        """输出到控制台"""
        print(formatted_message)
        sys.stdout.flush()
    
    def _write_to_file(self, formatted_message: str) -> None:
        """输出到文件"""
        if not self.file_output:
            return
        
        self._ensure_daily_log_file()
        
        if self.log_file:
            try:
                with open(self.log_file, 'a', encoding='utf-8') as f:
                    f.write(formatted_message + '\n')
            except Exception as e:
                print(f"写入日志文件失败: {e}")
    
    def log(self, message: str, level: LogLevel = LogLevel.INFO) -> None:
        """
        输出日志
        
        Args:
            message: 日志消息
            level: 日志级别
        """
        if not self._should_log(level):
            return
        
        formatted_message = self._format_message(message, level)
        
        if self.console_output:
            self._write_to_console(formatted_message)
        
        if self.file_output:
            self._write_to_file(formatted_message)
    
    def debug(self, message: str) -> None:
        """输出DEBUG级别日志"""
        self.log(message, LogLevel.DEBUG)
    
    def info(self, message: str) -> None:
        """输出INFO级别日志"""
        self.log(message, LogLevel.INFO)
    
    def warning(self, message: str) -> None:
        """输出WARNING级别日志"""
        self.log(message, LogLevel.WARNING)
    
    def error(self, message: str) -> None:
        """输出ERROR级别日志"""
        self.log(message, LogLevel.ERROR)
    
    def separator(self, char: str = "-", length: int = 70) -> None:
        """输出分隔线（不使用，保留兼容性）"""
        pass
    
    def section(self, title: str, char: str = "=", length: int = 70) -> None:
        """输出章节标题"""
        self.print_with_time(char * length)
        self.print_with_time(title)
        self.print_with_time(char * length)
    
    def print_with_time(self, message: str) -> None:
        """
        兼容旧的print_with_time函数
        输出带时间戳的消息（不带日志级别标签）
        """
        timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        formatted_message = f"[{timestamp}] {message}"
        
        if self.console_output:
            self._write_to_console(formatted_message)
        
        if self.file_output:
            self._write_to_file(formatted_message)
    
    def close(self) -> None:
        """关闭日志（预留接口，当前实现无需特殊清理）"""
        pass
    
    # ------------------------------------------------------------------
    # 内部工具方法
    # ------------------------------------------------------------------
    def _configure_log_directory(self, log_path: Optional[str]) -> None:
        """
        配置日志目录
        
        Args:
            log_path: 用户指定的日志路径（可以是文件或目录），None 表示使用当前工作目录
        """
        if log_path:
            abs_path = os.path.abspath(log_path)
            if os.path.isdir(abs_path):
                self._log_dir = abs_path
            else:
                self._log_dir = os.path.dirname(abs_path) or os.getcwd()
        else:
            self._log_dir = os.getcwd()
        
        os.makedirs(self._log_dir, exist_ok=True)
        # 初始化当前日志文件
        self._ensure_daily_log_file(force=True)
    
    def _ensure_daily_log_file(self, force: bool = False) -> None:
        """确保按天创建日志文件"""
        if not self.file_output or not self._log_dir:
            return
        
        today = datetime.now().strftime("%Y%m%d")
        if not force and self._current_day == today and self.log_file and os.path.exists(self.log_file):
            return
        
        self._current_day = today
        file_name = f"log_{today}.log"
        full_path = os.path.join(self._log_dir, file_name)
        
        # 如果文件不存在则创建；append_mode=False 时清空旧文件
        if not os.path.exists(full_path):
            open(full_path, 'w').close()
        elif not self.append_mode:
            open(full_path, 'w').close()
        
        self.log_file = full_path


# 全局默认日志实例（仅控制台输出）
_default_logger: Optional[Logger] = None


def get_default_logger() -> Logger:
    """获取默认日志实例"""
    global _default_logger
    if _default_logger is None:
        _default_logger = Logger(console_output=True, file_output=False)
    return _default_logger


def set_default_logger(logger: Logger) -> None:
    """设置默认日志实例"""
    global _default_logger
    _default_logger = logger


def print_with_time(message: str) -> None:
    """
    全局函数：输出带时间戳的消息
    使用默认日志实例
    """
    get_default_logger().print_with_time(message)


# 全局logger实例，供其他模块直接导入使用
logger = get_default_logger()

File: electron/logger.ts
import { app } from 'electron'
import { existsSync, mkdirSync, appendFileSync, writeFileSync } from 'fs'
import { join } from 'path'
import { tmpdir } from 'os'

/**
 * 日志级别
 */
export enum LogLevel {
  DEBUG = 0,
  INFO = 1,
  WARNING = 2,
  ERROR = 3,
}

/**
 * 日志管理器类
 * 参考 Python logger.py 的实现
 */
export class Logger {
  private logDir: string
  private currentDay: string | null = null
  private logFile: string | null = null
  private consoleOutput: boolean
  private fileOutput: boolean
  private logLevel: LogLevel
  private appendMode: boolean

  constructor(options: {
    logDir?: string
    consoleOutput?: boolean
    fileOutput?: boolean
    logLevel?: LogLevel
    appendMode?: boolean
  } = {}) {
    this.consoleOutput = options.consoleOutput ?? true
    this.fileOutput = options.fileOutput ?? true
    this.logLevel = options.logLevel ?? LogLevel.INFO
    this.appendMode = options.appendMode ?? true

    // 默认日志目录：根据环境选择
    if (options.logDir) {
      this.logDir = options.logDir
    } else {
      try {
        // 开发环境：使用项目根目录的 logs
        if (process.env.VITE_DEV_SERVER_URL) {
          const appPath = process.cwd()
          this.logDir = join(appPath, 'logs')
        } else {
          // 生产环境：使用 userData 目录（安装版和便携版都可写）
          // 延迟获取，避免在 app 未准备好时调用
          try {
            const userDataPath = app.getPath('userData')
            this.logDir = join(userDataPath, 'logs')
          } catch (appError) {
            // 如果 app 还未准备好，先使用临时目录，后续会重新初始化
            this.logDir = join(tmpdir(), 'yuntue-logs')
            console.warn('[Logger] app 未准备好，使用临时目录，将在 app ready 后重新初始化')
          }
        }
      } catch (error) {
        // 如果获取失败，使用临时目录
        console.error('[Logger] 无法获取日志目录，将禁用文件输出:', error)
        this.fileOutput = false
        this.logDir = ''
        return
      }
    }

    // 确保日志目录存在
    if (this.fileOutput) {
      this._ensureLogDirectory()
      
      // 初始化日志文件
      this._ensureDailyLogFile(true)
    }
  }

  /**
   * 格式化日志消息
   */
  private _formatMessage(message: string, level: LogLevel): string {
    const timestamp = new Date().toLocaleString('zh-CN', {
      year: 'numeric',
      month: '2-digit',
      day: '2-digit',
      hour: '2-digit',
      minute: '2-digit',
      second: '2-digit',
      hour12: false,
    })
    const levelName = LogLevel[level]
    return `[${timestamp}] [${levelName}] ${message}`
  }

  /**
   * 判断是否应该输出此级别的日志
   */
  private _shouldLog(level: LogLevel): boolean {
    return level >= this.logLevel
  }

  /**
   * 输出到控制台
   */
  private _writeToConsole(formattedMessage: string, level: LogLevel): void {
    if (!this.consoleOutput) return

    switch (level) {
      case LogLevel.ERROR:
        console.error(formattedMessage)
        break
      case LogLevel.WARNING:
        console.warn(formattedMessage)
        break
      case LogLevel.DEBUG:
        console.debug(formattedMessage)
        break
      default:
        console.log(formattedMessage)
    }
  }

  /**
   * 输出到文件
   */
  private _writeToFile(formattedMessage: string): void {
    if (!this.fileOutput) return

    this._ensureDailyLogFile()

    if (this.logFile) {
      try {
        appendFileSync(this.logFile, formattedMessage + '\n', 'utf-8')
      } catch (error) {
        console.error('写入日志文件失败:', error)
      }
    }
  }

  /**
   * 确保日志目录存在
   */
  private _ensureLogDirectory(): void {
    if (!this.logDir) {
      this.fileOutput = false
      return
    }

    if (!existsSync(this.logDir)) {
      try {
        mkdirSync(this.logDir, { recursive: true })
        console.log('[Logger] 创建日志目录:', this.logDir)
      } catch (error) {
        console.error('[Logger] 创建日志目录失败，禁用文件输出:', error)
        this.fileOutput = false
      }
    }
  }

  /**
   * 确保按天创建日志文件
   */
  private _ensureDailyLogFile(force: boolean = false): void {
    if (!this.fileOutput) return

    const today = new Date().toISOString().slice(0, 10).replace(/-/g, '')
    
    if (!force && this.currentDay === today && this.logFile && existsSync(this.logFile)) {
      return
    }

    this.currentDay = today
    const fileName = `main-${today}.log`
    this.logFile = join(this.logDir, fileName)

    // 如果文件不存在则创建；appendMode=false 时清空旧文件
    try {
      // 确保目录存在
      this._ensureLogDirectory()
      
      if (!existsSync(this.logFile)) {
        writeFileSync(this.logFile, '', 'utf-8')
      } else if (!this.appendMode) {
        writeFileSync(this.logFile, '', 'utf-8')
      }
    } catch (error) {
      console.error('[Logger] 无法创建日志文件，禁用文件输出:', error)
      this.fileOutput = false // 禁用文件输出以避免后续错误
    }
  }

  /**
   * 输出日志
   */
  log(message: string, level: LogLevel = LogLevel.INFO): void {
    if (!this._shouldLog(level)) return

    const formattedMessage = this._formatMessage(message, level)

    this._writeToConsole(formattedMessage, level)
    this._writeToFile(formattedMessage)
  }

  /**
   * 输出 DEBUG 级别日志
   */
  debug(message: string): void {
    this.log(message, LogLevel.DEBUG)
  }

  /**
   * 输出 INFO 级别日志
   */
  info(message: string): void {
    this.log(message, LogLevel.INFO)
  }

  /**
   * 输出 WARNING 级别日志
   */
  warning(message: string): void {
    this.log(message, LogLevel.WARNING)
  }

  /**
   * 输出 ERROR 级别日志
   */
  error(message: string): void {
    this.log(message, LogLevel.ERROR)
  }

  /**
   * 获取日志目录路径
   */
  getLogDir(): string {
    return this.logDir
  }

  /**
   * 获取当前日志文件路径
   */
  getCurrentLogFile(): string | null {
    return this.logFile
  }
}

// 全局默认日志实例
let _defaultLogger: Logger | null = null

/**
 * 获取默认日志实例
 */
export function getDefaultLogger(): Logger {
  if (!_defaultLogger) {
    try {
      _defaultLogger = new Logger({
        consoleOutput: true,
        fileOutput: true,
        logLevel: LogLevel.INFO,
        appendMode: true,
      })
    } catch (error) {
      // 如果初始化失败（例如 app 还未准备好），创建一个仅控制台输出的 logger
      console.error('[Logger] 初始化失败，使用仅控制台输出模式:', error)
      _defaultLogger = new Logger({
        consoleOutput: true,
        fileOutput: false,
        logLevel: LogLevel.INFO,
        appendMode: true,
      })
    }
  }
  return _defaultLogger
}

/**
 * 设置默认日志实例
 */
export function setDefaultLogger(logger: Logger): void {
  _defaultLogger = logger
}

// 导出全局日志实例（使用 getter，延迟初始化）
let _loggerInstance: Logger | null = null

function getLogger(): Logger {
  if (!_loggerInstance) {
    try {
      _loggerInstance = getDefaultLogger()
    } catch (error) {
      // 如果获取失败，返回一个临时的仅控制台 logger
      _loggerInstance = new Logger({
        consoleOutput: true,
        fileOutput: false,
        logLevel: LogLevel.INFO,
      })
    }
  }
  return _loggerInstance
}

// 导出 logger 对象，使用 getter 属性
export const logger = {
  info: (msg: string) => getLogger().info(msg),
  error: (msg: string) => getLogger().error(msg),
  warning: (msg: string) => getLogger().warning(msg),
  debug: (msg: string) => getLogger().debug(msg),
  log: (msg: string, level?: LogLevel) => getLogger().log(msg, level),
}

